### DISCUSSION TITLE
Need Guidance regarding GSOC 2021

### DISCUSSION THREAD

**Discussion by Hetarth02** [networkx/networkx#4666] [Q&A] [answered]
Hello, this my first time contributing to an open-source project and I want take part in GSOC 2021 and do a project under NetworkX. Any guidance on how to do so is appreciated.

**Comment by MridulS** [comment, marked as answer]
That's great to hear @Hetarth02!

As a first step you could go through the ideas list http://networkx.org/documentation/latest/developer/projects.html#mentored-projects and see if you are interested in any of them. (You can also propose your own idea for a NetworkX project!)

To get acquainted with our development process please go through https://networkx.org/documentation/latest/developer/contribute.html#dev-workflow to setup
your local dev environment and https://networkx.org/documentation/latest/developer/new_contributor_faq.html for
some common FAQs.

For some issues to work on to get familiarised with NetworkX you can look at
 https://github.com/networkx/networkx/issues?q=is%3Aissue+is%3Aopen+label%3ASprint or you can pick any
other issue on the GitHub issue tracker https://github.com/networkx/networkx/issues.

If you interested in the notebooks project you can also look at https://github.com/networkx/notebooks/issues/13

**Reply by Hetarth02** [reply]
Hello @MridulS and Thank you for responding to me.
I am particularly interested in,
"Pedagogical Interactive Notebooks for Algorithms Implemented in NetworkX"
under the mentored projects section.

I think I may be able to help in the documentation of some of the algorithms or methods listed there. But currently, I do not know how to, where to or whom to sent my application. Can you help me solve any of the above queries?

**Reply by MridulS** [reply]
If you want to apply for GSoC, please first go through the GSoC website https://summerofcode.withgoogle.com they have everything laid out.

There is an official student guide by Google https://google.github.io/gsocguides/student/ which should also answer all of your questions :)

Let us know if you have any questions regarding it.

**Reply by Hetarth02** [reply]
@MridulS I read the student guide and now some of the topics are cleared. But I have some questions regarding interaction with the mentors and the community.

1.) Do I need to interact with the community and my mentors at the git thread (https://github.com/networkx/notebooks) which is related to my interested project only?

2.) Is it mandatory for me to contribute in any kind of way atleast once before GSOC 2021?

3.) If I want to contribute out of GSOC but with the list of topics given in GSOC then, can I? If so, do I only need to PR my ipynb file?

**Reply by MridulS** [reply]
>  Do I need to interact with the community and my mentors at the git thread (https://github.com/networkx/notebooks) which is related to my interested project only?

You can use this discussion thread if you have any general questions! If you want to discuss about something specific (like a particular algorithm that you would like to implement in a notebook) you can open an issue in the https://github.com/networkx/notebooks repo. 

>  Is it mandatory for me to contribute in any kind of way atleast once before GSOC 2021?

It will definitely help your application!

>  If I want to contribute out of GSOC but with the list of topics given in GSOC then, can I? If so, do I only need to PR my ipynb file?

Yes! NetworkX is an open source project and you can always contribute a notebook (ipynb) file to the notebooks repo irrespective of GSoC.

**Reply by Hetarth02** [reply]
@MridulS Thank you for your help. Looking forward to participate and contribute.

**Comment by Hetarth02** [comment]
@MridulS As you suggested earlier,
Here is the link to my draft proposal.
[Draft_Proposal](https://drive.google.com/file/d/1VV91hdlKCsERi3b9YjxO7yFckwk2_-OE/view?usp=drivesdk)
I have given the comment rights, so do suggest me if there is anything missing or if something is wrong.

Thanks in advance.

**Reply by MridulS** [reply]
Thanks for this @Hetarth02

Just some quick comments:
- In my opinion you are trying to cover *a lot* of areas, producing 2-3 detailed notebooks should be an achievable goal for a 175 hour GSoC.

- We don’t want the notebooks gallery to be a copy of the documentation. The plan is to have narrative notebooks with detailed and intuitive explanations of the algorithms.
  - An example could be a notebook on clustering
	- What does it means? Mathematically and intuitively.
	- Different kinds of clustering algorithms ( in networkx and not in networkx)
	- Pick up https://networkx.org/documentation/latest/reference/algorithms/generated/networkx.algorithms.cluster.clustering.html and explain the algorithm with examples.
  - Or a notebook on PageRank
    - Maybe a bit about its history
    - How does it differ from other centrality measures/link analysis?
    - The idea of eigenvectors in link analysis.
    - An example to follow along with the algorithm.

Let me know if you have any other questions.

Mridul

PS: these are just examples, feel free to pick up anything that fancies you :)

**Reply by Hetarth02** [reply]
@MridulS I think I get the idea but correct me if I don't, you need a notebook which explains mathematical history of the topic, explaining the algorithms, basically an in-depth guide right?

I appreciate your feedback on the proposal, I will update my proposal according to it. But can you check these 2 notebooks I made earlier [Notebooks](https://github.com/networkx/notebooks/pull/19) and tell me what else do I need to add to it or how can I improve it? It would really help me clear some doubts.

And if possible can you update me as soon as you can?

Thanks in advance.
